src/main/kotlin/org/jetbrains/research/ml/codetracker/server/CollectionsQueryExecutor.kt

```diff
@@ -19,7 +19,22 @@ object CollectionsQueryExecutor : QueryExecutor() {
     }
 
     inline fun <reified T : Any> getCollection(url: String, serializer: KSerializer<T>): List<T> {
-        return parseResponse(
+        return parseCollectionResponse(
+            executeQuery(
+                Request.Builder().url(URL("${baseUrl}${url}")).build()
+            ), serializer
+        )
+    }
+
+    inline fun <reified T : Any> parseItemResponse(response: Response?, serializer: KSerializer<T>): T {
+        if (isSuccess(response)) {
+            return Json(JsonConfiguration.Stable).parse(serializer, response?.body?.string() ?: "")
+        }
+        throw IllegalStateException("Unsuccessful server response")
```

Comment:
These lines are repeated twice (here and at lines 15-19), let's also move them out?
for example, 
```
    private fun <T, R> parseResponse(response: Response?, serializer: KSerializer<T>, transform: (KSerializer<T>) -> KSerializer<R>): R {
        if (isSuccess(response)) {
            return Json(JsonConfiguration.Stable).parse(transform(serializer), response?.body?.string() ?: "")
        }
        throw IllegalStateException("Unsuccessful server response")
    }
```

Sooo considering the previous refactoring it all can look like this (you can replace `T : Any` with just `T` and get rid of `inline` and `retifed`):
```
object CollectionsQueryExecutor : QueryExecutor() {

    fun <T> getCollection(url: String, serializer: KSerializer<T>): List<T> {
        return parseResponse(getResponse(url), serializer) { it.list }
    }

    fun <T> getItemFromCollection(url: String, serializer: KSerializer<T>): T {
        return parseResponse(getResponse(url), serializer) { it }
    }

    private fun <T, R> parseResponse(response: Response?, serializer: KSerializer<T>, transform: (KSerializer<T>) -> KSerializer<R>): R {
        if (isSuccess(response)) {
            return Json(JsonConfiguration.Stable).parse(transform(serializer), response?.body?.string() ?: "")
        }
        throw IllegalStateException("Unsuccessful server response")
    }

    private fun getResponse(url: String): Response? {
        return executeQuery(Request.Builder().url(URL("${baseUrl}${url}")).build())
    }
}
```

Reply:
It looks better, thanks! I will refactor it